edb/server/server.py

```diff
@@ -754,23 +760,56 @@ async def introspect_db_config(self, conn):
         result = await conn.sql_fetch_val(self.get_sys_query('dbconfig'))
         return config.from_json(config.get_settings(), result)
 
-    async def _introspect_dbs(self):
-        syscon = await self._acquire_sys_pgcon()
+    async def _early_introspect_db(self, dbname):
+        """We need to always introspect the extensions for each database.
+
+        Otherwise we won't know to accept connections for graphql or
+        http, for example, until a native connection is made.
+        """
+        logger.info("introspecting extensions for database '%s'", dbname)
+
+        conn = await self._acquire_intro_pgcon(dbname)
+        if not conn:
+            return
+
         try:
-            dbs_query = self.get_sys_query('listdbs')
-            json_data = await syscon.sql_fetch_val(dbs_query)
-            dbnames = json.loads(json_data)
-        finally:
-            self._release_sys_pgcon()
+            extension_names_json = await conn.sql_fetch_val(
+                b'''
+                    SELECT json_agg(name) FROM edgedb."_SchemaExtension";
```

Comment:
I'd really prefer this to be a properly compiled EdgeQL "sys_query" like the other introspection queries here.